accepted/2022/simplify-output-paths.md

```diff
@@ -0,0 +1,77 @@
+# Simplify .NET Output Paths
+
+We'd like to simplify the output paths for .NET 8.
+
+Currently, the default output path for a .NET project includes folders for the `TargetFramework` and the `RuntimeIdentifier` (if it is set).  Additionally, the publish output goes in a `publish` subfolder of the output folder.  The resulting paths are as follows:
+
+- `bin\<Configuration>\<TargetFramework>` - Build output with no `RuntimeIdentifier`
+- `bin\<Configuration>\<TargetFramework>\<RuntimeIdentifier>` - Build output with `RuntimeIdentifier`
+- `bin\<Configuration>\<TargetFramework>\publish` - Publish output with no `RuntimeIdentifier`
+- `bin\<Configuration>\<TargetFramework>\<RuntimeIdentifier>\publish` - Publish output with `RuntimeIdentifier`
+
+This is rather complicated and inconsistent, which can make for a poor first impression of the .NET platform.
+
+We'd like to try to improve the output path structure.  Desired qualities include:
+
+- Simple and consistent
+- Avoid cases where one output path is nested inside of another one
+- Avoid excessively deep directory hierarchies
+- Get rid of `obj` folder in project root
+
+## Proposed behavior
+
+The new output path format will consist of the following 3 nested folders:
+
+- `artifacts` - All output will go under this folder
```

Comment:
Did we think about providing an easier way to move the artifacts folder to the root of the repo?
e.g. all our repos use `<root>\artifacts\bin\<project name>\` today via some arcade settings: https://github.com/dotnet/arcade/blob/7cd5e97416ad782621a2cce154d659253c06223a/src/Microsoft.DotNet.Arcade.Sdk/tools/ProjectLayout.props

If we had a way to set a single property to define the root of the artifacts path we could remove all that customization.

Reply:
I'm not sure I actually like this idea, but have we looked at how other IDE's like Xcode do things?  By default they have an output path for build items that's more 'global' on the machine: `~/Library/Developer/Xcode/DerivedData/[PROJECT]/...`

This moves the data out of the project folder entirely by default and would avoid accidental globbing of things in the folder, and at some level helps keep MAX_PATH limitations more predictable.

You can of course override the default path with xcodebuilds too, and there is the issue of discoverability of the location of these things, but maybe worth musing on a bit?

Reply:
How does this work when I have multiple projects (think TestProject) all over my machine? Is there some sha/guid/id that is used to distinguish between them?

Reply:
I think we should consider putting this folder in the solution root by default. That is true even when building individual projects. Much like editor config support we can "look up" until we find a solution and put the artifact folder there. 

That would put us on the course of keeping source directories clean (read only) during builds and move all the writes to a single known directory.

Reply:
What exactly are the rules for this? _Looking up_ by convention isn't a bad thing, but it's not entirely reliable either.

1. What happens if there nested solution files (`*.sln`)? You can stop, but you're not at the root.
   a. What about a solution filter (`*.slnf`)? I would presume it should use the same rules.
2. What if there is no solution file (`*.sln`)? That's not strictly required to build.

There are other sensible conventions too, such as the presence of the `/.git` folder or `.gitignore`, but you don't have to use Git nor is a codebase necessarily under source control.

_Convention over Configuration_ should be preferred 💯 , but there should be an _escape hatch_ since there is no guarantee the configuration can otherwise be met. For example, one approach might be to have a zero-byte file `Build.Root` indicate the what the root directory is by the presence of the file if it can't otherwise be determined. On one hand, this is rife for abuse and can fall into the pitfall as a other possibly nested files such as solutions. On the other hand, if you know about this capability and are using it, you understand the minutia of how and why - so caveat emptor. Such an approach would be consistent with how `.editorconfig`, `nuget.config`, and other hierarchical resolutions work.

At a minimum, it needs to be clearly defined and documented as to what happens if root folder cannot be determined from any known convention.

Reply:
> What happens if there nested solution files (*.sln)? You can stop, but you're not at the root.

Similar to what happens if you have nested `.editorconfig` files. The first one found wins. 

> What if there is no solution file (*.sln)? That's not strictly required to build.

In that case put the folder next to the project file. 

Another potential convention to look for is the presence of a `Directory.Build.props` file. These generally a root point of a build and are often next to solution files. Unlike solution files they also have the mechanism to invent convention by adding a property the build could process for where to put the artifacts folder.

> At a minimum, it needs to be clearly defined and documented as to what happens if root folder cannot be determined from any known convention.

Agree. Have to be clearer about this.

Reply:
While I like the idea of having solution artifacts folder, I feel like building is already too slow. I'm concerned all these checks (search for *.sln files, search for Directory.Build.props, search for Directory.Build.targets, search for Directory.Packages.props) is just adding additional perf paper-cuts.

As it is, with NuGet.Client, if I run `msbuild -restore -t:Build -v:m -m` twice, the second one, when everything is already up-to-date, takes 46 seconds, which I feel is too slow. Admittedly when I change 1 line of code in a test project and run `dotnet test` it doesn't need to check all ~90 projects in the solution, but still I don't want to wait 30+ seconds for project references to be validated before my test project is compiled and finally tests run.

I recently helped a customer who claimed their repo contains 18,000 projects. I still can't get my head around that. I know there are several repos within Microsoft with 1000+ projects.  All this is just a rambly (ranting?) way of saying I think build performance should also be a consideration.

Reply:
Fair point, which is why there should be an explicitly, configurable _escape hatch_ that configures the root. `Directory.Build.props` isn't going away anytime soon. That is the place you'd likely explicitly configure the root path, which would supplant resolution by convention. There definitely needs to be a fallback if no convention matches, but an explicit option is also beneficial for scenarios (ex: large projects) that aren't in the 80% and need to squeeze every last drop of build performance.

Reply:
> I think we should consider putting this folder in the solution root by default. That is true even when building individual projects

This is a good discussion.

How would we handle building individual projects, that may not have a .sln or be part of a git repo.  What would this look like for someone who is getting started in .NET and just runs `dotnet new console`?

Reply:
> How would we handle building individual projects, that may not have a .sln or be part of a git repo

I think we fall back on the conventions we established with .editorconfig. Essentially look up until we find a valid solution. If none is found then we fall back to the current directory. 

> What would this look like for someone who is getting started in .NET and just runs dotnet new console?

In this case there is no solution here, presumably up the directory parent path too, so the `artifacts` folder ends up in the current directory. 

I also think we should consider `.artifacts` instead of `artifacts`. The reasoning being what value does this have to the average user of .NET? Commands like `build`, `run`, `pack` and `publish` don't require users to understand how we structure our build artifacts. That is of use to a minority of .NET customers. As such I think we should consider a `.` prefixed name to _hide_ it in many views. 

For users who do actually look at build outputs this change doesn't impede them in any way. The hiding is by convention only. Knowing the name is enough to process them the same way they do today.

Reply:
> I also think we should consider `.artifacts` instead of `artifacts`.

Great suggestion!  This actually solves a lot of the problems with having a new named folder in the project folder, as folders that start with `.` are already excluded from the default globs.

Reply:
I'm not sure the `.artifacts` is a good change. If I build a MAUI app on macOS on the CLI today and then try to run it by double clicking on the app I'd need to configure Finder to show hidden folders which requires knowing the special keyboard combination `CMD+Shift+.`

Reply:
Developres would be able to configure the name, this is about what the default would be. I think the default should favor the majority use and the overwhelming majority use is it's adding noise to the customer experience. This is particularly bad for new users to the platform as it increases the concept count they have to handle.

Reply:
With response to the perf concerns, I don't think that will be an issue with the updated proposal.  The proposed logic is to first look for a `Directory.Build.props` file to find where the `.artifacts` folder should go.  Since MSBuild already walks up the folder tree looking for this file, we should be able to use that folder (if it exists) without an additional directory scan.

The proposal does specify that if there isn't a `Directory.Build.props` file then we'll walk up the tree looking for a `.sln` folder.  However, this would only happen if the new output format was enabled, which is typically done in `Directory.Build.props`.  The `.sln` logic is mainly there to make it easier to turn on the new output format by default if we ever do that.